Rust cargo verify

fshaked opened the pull request:
Rewrite of cargo-verify in rust.
All the features were copied from the python script, but not very well tested.
Some of the command line arguments were changed, see --help.

fshaked committed:
(big commit) Rewrite of cargo-verify in rust.

alastairreid commented on cargo-verify/Cargo.toml:
```diff
@@ -0,0 +1,21 @@
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+structopt = "0.3"
```
Can you sort these alphabetically?
(Makes it more obvious where to add new dependencies and matches the 'use' lines)

alastairreid commented on cargo-verify/Cargo.toml:
```diff
@@ -0,0 +1,21 @@
+num_cpus = "1.0"
+glob = "0.3.0"
+shell-escape = "0.1.5"
+# inkwell = { git = "https://github.com/alastairreid/inkwell", features = ["llvm10-0"] }
```
Remove this?

alastairreid commented on cargo-verify/src/utils.rs:
```diff
@@ -0,0 +1,134 @@
+
+use log::info;
+
+// `info_at!(&opt, level, ...)` will print the formatted message `...` if
```
Should this and other comments use /// instead of // so that we could generate documentation?

alastairreid commented on cargo-verify/src/utils.rs:
```diff
@@ -0,0 +1,134 @@
+// The Append trait lets you chain `append` calls where usually you would have
+// to mutate (e.g. using `push`).
+// Example:
+// assert_eq!(String::from("foo").append("bar"), { let mut x = String::from("foo"); x.push_str("bar"); x })
```
Should this be a doctest?

fshaked replied:
Probably, but we will have to wrap the code in a library.

alastairreid commented on cargo-verify/src/utils.rs:
```diff
@@ -0,0 +1,134 @@
+}
+
+/// Example:
+/// assert_eq!(add_pre_ext(&PathBuf::from("foo.bar"), "baz"), PathBuf::from("foo.baz.bar"))
```
doctest?

alastairreid commented on cargo-verify/src/utils.rs:
```diff
@@ -0,0 +1,134 @@
+    }
+}
+
+/// Example:
```
Should this comment describe the function in words?

alastairreid commented on cargo-verify/src/proptest.rs:
```diff
@@ -0,0 +1,61 @@
+    let stdout = from_utf8(&output.stdout).expect("stdout is not in UTF-8");
+    let stderr = from_utf8(&output.stderr).expect("stderr is not in UTF-8");
+
+    if !output.status.success() {
```
Not critical to do now - but we should plan on doing a cleanup pass that factors out this error-handling then-branch so that it is shared with more functions.
Probably everything from line 41 (info_cmd) down to 57 (Ok(Status::Error)).

fshaked replied:
Done.

alastairreid commented on cargo-verify/src/seahorn.rs:
```diff
@@ -0,0 +1,182 @@
+// 1: brief description of error
+// 2: long details about an error
+// 3: warnings
+// 4: non-KLEE output
```
Change KLEE to SeaHorn?

alastairreid commented on cargo-verify/src/main.rs:
```diff
@@ -0,0 +1,841 @@
+
+#[derive(Debug, PartialEq, Copy, Clone)]
+pub enum Status {
+    Unknown, // E.g. the varifier failed to execute.
```
verifier not varifier

alastairreid commented on cargo-verify/src/main.rs:
```diff
@@ -0,0 +1,841 @@
+pub enum Status {
+    Unknown, // E.g. the varifier failed to execute.
+    Verified,
+    Error, // E.g. the varifier found a violation.
```
verifier not varifier

alastairreid commented on cargo-verify/src/main.rs:
```diff
@@ -0,0 +1,841 @@
+    Ok(())
+}
+
+// Compile a Rust crate to generate bitcode
```
Should this (and other comments in this file and other files) use /// instead of //?

alastairreid commented on cargo-verify/src/main.rs:
```diff
@@ -0,0 +1,841 @@
+    } else {
+        vec![("main".to_string(), "main".to_string())]
+    };
+    // Remove the package name from the function names (important for Klee?).
```
I couldn't figure out which part of cargo-verify this corresponds to

fshaked replied:
The line below the comment :)
The problem starts with the name "tests" in the code, what it should really be is "functions", because in general we check functions (at the moment it can be `main` or one of the test functions).
Now, what the comment is trying to say is that after we get the mapping (`tests` maps function names to mangled names), we remove from each key the prefix `<package-name>::`, if the key has such prefix, because, I think, when we pass this name to Klee it expects it to be without the prefix.

alastairreid commented on cargo-verify/src/main.rs:
```diff
@@ -0,0 +1,841 @@
+    // The following line is not present because we care about the target It is
+    // there to allow us to use -Clto to build crates whose dependencies invoke
+    // proc_macros.
+    // FIXME: "=="?
```
This can only get fixed once cargo/rustc becomes less crazy.
Until then, we need this workaround.

fshaked replied:
I don't understand what is going on here. We are passing `--target=={target}` to `cargo build`? What is the effect of this?
I changed this `==` to `=`, should I change it back?

alastairreid replied:
I think my comment was a response to an older? comment in the code about the whole idea of using --target, (not = vs ==).

= is the right thing to use - == must be a typo by me that cargo fails to detect as an error

alastairreid commented on cargo-verify/src/main.rs:
```diff
@@ -0,0 +1,841 @@
+    // {target_dir}/{target}/debug/deps/{package}*.bc
+    let bc_files = glob(
+        &glob::Pattern::escape(
+            target_dir
```
Would this chain of appends not be clearer as [target_dir, target, "debug", "deps", package]?
or as format!("{}/{}/debug/deps/{}*.bc", target_dir, target, package)?

fshaked replied:
target_dir is not printable (need to use `target_dir.to_string_lossy()`); I also need to pass everything to `escape`, incase they use something that might be interpreted by glob as something else, but I must not pass the `*` through escape. So I don't think there is something significantly better that I can do here.

Initially I did this without using the globe crate and it was amazingly ugly.

alastairreid commented on cargo-verify/src/main.rs:
```diff
@@ -0,0 +1,841 @@
+    };
+
+    // {targetdir}/{target}/debug/build/ * /out/ *.o"
+    let c_files = glob(
```
see comment on previous glob

alastairreid commented on cargo-verify/src/main.rs:
```diff
@@ -0,0 +1,841 @@
+    let mut cmd = Command::new("llvm-link");
+    cmd.arg("-o").arg(out_file).args(in_files);
+
+    utils::info_cmd(&cmd, "llvm-link");
```
This pattern of error handling occurs a lot

alastairreid commented on cargo-verify/src/main.rs:
```diff
@@ -0,0 +1,841 @@
+    Ok((bc_file, c_files))
+}
+
+// Link multiple bitcode files together.
```
I think this should be in common_backend.rs or utils.rs or similar.
All it does is call an existing utility program in a very generic way.

fshaked replied:
It is now just one line (actually cargo fmt broke it over multiple lines) in it's inlined (no function).

alastairreid commented on cargo-verify/src/main.rs:
```diff
@@ -0,0 +1,841 @@
+    Ok(())
+}
+
+// Patch LLVM file to enable verification
```
Use /// instead of //?

alastairreid commented on cargo-verify/src/main.rs:
```diff
@@ -0,0 +1,841 @@
+//   (this makes std::env::args() work)
+// - redirecting panic! to invoke backend-specific intrinsic functions
+//   for reporting errors
+fn patch_llvm(opt: &Opt, options: &[&str], bcfile: &Path, new_bcfile: &Path) -> CVResult<()> {
```
Move to utils.rs or similar?

alastairreid commented on cargo-verify/src/main.rs:
```diff
@@ -0,0 +1,841 @@
+    Ok(())
+}
+
+fn get_meta_package_name(opt: &Opt) -> CVResult<String> {
```
Add /// comment.

Move to utils.rs?

alastairreid commented on cargo-verify/src/main.rs:
```diff
@@ -0,0 +1,841 @@
+    Ok(name)
+}
+
+fn get_meta_target_directory(opt: &Opt) -> CVResult<PathBuf> {
```
Add ///
Move to utils.rs?

alastairreid commented on cargo-verify/src/main.rs:
```diff
@@ -0,0 +1,841 @@
+    Ok(dir)
+}
+
+fn get_default_host(crate_path: &Path) -> CVResult<String> {
```
Add ///
Move to utils.rs?

alastairreid commented on cargo-verify/src/main.rs:
```diff
@@ -0,0 +1,841 @@
+}
+
+// Generate a list of tests in the crate by parsing the output of `cargo test --
+// --list`
```
Unfortunate line break in middle of command.
Consider splitting line earlier?

alastairreid commented on cargo-verify/src/main.rs:
```diff
@@ -0,0 +1,841 @@
+    Ok(tests)
+}
+
+// Find a function defined in LLVM bitcode file
```
The comment for the function with the same name in Python script talks about what happens with the hash at the end of the mangled name.

More broadly, I think it would be best to copy the comments from the Python script exactly as they are and then improve them instead of deleting them or writing something else.

fshaked replied:
Sorry about the comments. I mostly tried to do just that, at least in my head I thought I was doing that :).

alastairreid reviewed (commented):
A few general things that I commented on a few times but that apply in many places in the code.

- use /// comments before functions and add a comment before every function
  - if the Python code had a comment before a function, copy that over and improve/update it as appropriate.
- There are a bunch of functions that just execute a program and report errors. I think these should be in utils.rs or backends_common.rs unless they do something interesting.
- A lot of code that executes an external program has a similar pattern of code to report the error - roughly starting from "info_cmd" and continuing down until the error has been reported. It would be good to factor that out into a macro or helper function.

fshaked commented:
>   * if the Python code had a comment before a function, copy that over and improve/update it as appropriate.

I mostly did that, I might have missed one or two, I will check.

> * There are a bunch of functions that just execute a program and report errors. I think these should be in utils.rs or backends_common.rs unless they do something interesting.

I disagree, those are cargo-verify's core functionality; I think utils.rs is for things we wish were in std.

> * A lot of code that executes an external program has a similar pattern of code to report the error - roughly starting from "info_cmd" and continuing down until the error has been reported. It would be good to factor that out into a macro or helper function.

Ok.

alastairreid commented:
> > * There are a bunch of functions that just execute a program and report errors. I think these should be in utils.rs or backends_common.rs unless they do something interesting.
> 
> I disagree, those are cargo-verify's core functionality; I think utils.rs is for things we wish were in std.

I don't think they should be in main.rs.  I don't care if that is one of the two files I mentioned or a new file.
But I think they are cluttering up main.rs.

fshaked commented:
> > > * There are a bunch of functions that just execute a program and report errors. I think these should be in utils.rs or backends_common.rs unless they do something interesting.
> > 
> > 
> > I disagree, those are cargo-verify's core functionality; I think utils.rs is for things we wish were in std.
> 
> I don't think they should be in main.rs. I don't care if that is one of the two files I mentioned or a new file.
> But I think they are cluttering up main.rs.

Ok

fshaked commented:
I fixed almost everything mentioned above. I didn't fix the doctests because it needs a workaround to work with bin crates, and I didn't fix the complex use of glob because I don't think it's possible (see replies above).

I also incorporated most of the small commits @alastairreid did to the rcv branch, some by using `git cherry-pick` and some by hand. I didn't change Klee's output dir back to `kleeout-{test}` (2ea8598) because that will litter the crate dir if you have a lot of tests. And instead of renaming the files we generate (f602cf6) I filter the .bc files looking for exactly one '.'.

alastairreid commented on cargo-verify/src/main.rs:
```diff
@@ -0,0 +1,677 @@
+        .append("*.bc"),
+    )?
+    .filter_map(Result::ok)
+    // Filter only files that have exactly one '.'
```
Can you make this comment better?
I think you are doing this to distinguish between files created by Cargo and files created by us (from patch or link) - but I worry that we will forget that really important detail!

alastairreid reviewed (commented):
If you can improve that one comment about the really subtle filenaming choice/filtering,  I'll approve this.

alastairreid reviewed (approved):
LGTM
